src/main/scala/async/Listener.scala

```diff
@@ -1,73 +1,178 @@
 
-  // TODO this should only be used for adapting the source
-  inline def mappingTopLock(inner: TopLock | Null, inline body: (Async.Source[?]) => LockResult) =
-    if inner != null then
-      new TopLock:
-        val selfNumber = inner.selfNumber
-        def lockSelf(source: Source[?]) = body(source)
+    /** Attempt to release all locks up to and including the given [[Listener.LockMarker]]. */
+    @tailrec
+    final def releaseAll(to: Listener.LockMarker): Unit =
+      val rest = release(to)
+      if rest != null then rest.releaseAll(to)
 
-  def acceptingListener[T](consumer: T => Unit) =
-    new Listener[T]:
-      val topLock = null
-      def complete(data: T, source: Source[T]) = consumer(data)
-      def release(to: ReleaseBoundary) = null
+    /** Attempt to lock all layers of this listener lock. */
+    private[Listener] final def lockAll(source: Async.Source[?]): Locked.type | Gone.type =
+      lockSelf(source) match
+        case Locked => Locked
+        case Gone => Gone
+        case inner: PartialLock => lockRecursively(inner)
 
-trait Listener[-T]:
-  val topLock: Listener.TopLock | Null
+    @tailrec
+    private final def lockRecursively(l: Listener.PartialLock): Locked.type | Gone.type =
+      l.lockNext() match
+        case Locked => Locked
+        case Gone =>
+          this.releaseAll(l)
+          Gone
+        case inner: PartialLock => lockRecursively(inner)
+  end ListenerLock
 
-  def complete(data: T, source: Async.Source[T]): Unit
-  def release(to: Listener.ReleaseBoundary): Listener[?] | Null
-
-  @tailrec
-  final def releaseAll(until: Listener.ReleaseBoundary): Unit =
-    val rest = release(until)
-    if rest != null then rest.releaseAll(until)
-
-  @tailrec
-  private def lock(l: Listener.SemiLock): Locked.type | Failed.type =
-    l.lockNext() match
-      case Locked => Locked
-      case Failed =>
-        this.releaseAll(l)
-        Failed
-      case inner: SemiLock => lock(inner)
-
-  def lockCompletely(source: Async.Source[T]): Locked.type | Failed.type =
-    this.topLock match
-      case topLock: Listener.TopLock =>
-        topLock.lockSelf(source) match
-          case Locked => Locked
-          case Failed => Failed
-          case inner: SemiLock => lock(inner)
-      case _ => Locked
+  /** A special wrapper for [[ListenerLock]] that just passes the source through. */
+  class ListenerLockWrapper(inner: ListenerLock, src: Async.Source[?]) extends ListenerLock:
+    val selfNumber: Long = inner.selfNumber
+    def lockSelf(_src: Async.Source[?]) =
+      inner.lockSelf(src)
+    def release(to: LockMarker): ListenerLock | Null = inner
 
-  def completeNow(data: T, source: Async.Source[T]): Boolean =
-    lockCompletely(source) match
-      case Locked =>
-        this.complete(data, source)
-        true
-      case Failed => false
+  /** Maps the lock of a listener, if it exists. */
+  inline def withLock[T](listener: Listener[?])(inline body: ListenerLock => T): T | Null =
```

Comment:
I'm not totally sure, but can't we declare the listener as inline as well? We only use it once, so this would avoid a possible variable declaration - but maybe this is optimised anyway.
```suggestion
  inline def withLock[T](inline listener: Listener[?])(inline body: ListenerLock => T): T | Null =
```

Reply:
It probably will ^^